pkg/cached/cached.go

```diff
@@ -2,368 +2,598 @@ package cached
+	if err := exec.Run(ctx, "vgextend", append([]string{"--config=devices/allow_mixed_block_sizes=1", c.VG}, c.ThinpoolPVs...)...); err != nil {
+		return fmt.Errorf("failed to extend vg with thinpool pvs: %w", err)
+	}
 
-func first(ss ...string) string {
-	for _, s := range ss {
-		if s != "" {
-			return s
-		}
+	if err := lvm.EnsureLV(ctx, c.ThinpoolLV, LVCreateThinpoolArgs(c.VG, c.ThinpoolPVs)...); err != nil {
+		return err
 	}
-	return ""
-}
 
-func getFolderSize(path string) (int64, error) {
-	var totalSize int64
-	err := filepath.Walk(path, func(_ string, info os.FileInfo, err error) error {
-		if err != nil {
+	if c.ThinpoolCacheLVSizeKib != "" {
+		if _, err := os.Stat(c.ThinpoolCachePV); err == nil {
+			logger.Info(ctx, "ram disk already exists", key.Device.Field(c.ThinpoolCachePV))
+		} else {
+			logger.Info(ctx, "creating ram disk", key.Device.Field(c.ThinpoolCachePV))
+			if err := exec.Run(ctx, "modprobe", "brd", "rd_nr=1", "rd_size="+c.ThinpoolCacheLVSizeKib); err != nil {
+				return fmt.Errorf("failed to create ram disk: %w", err)
+			}
+		}
+
+		if err := lvm.EnsurePV(ctx, c.ThinpoolCachePV); err != nil {
 			return err
 		}
-		if !info.IsDir() {
-			totalSize += info.Size()
+
+		if err := exec.Run(ctx, "vgextend", c.VG, c.ThinpoolCachePV); err != nil {
+			return fmt.Errorf("failed to extend vg with ram disk: %w", err)
+		}
+
+		if err := exec.Run(ctx, "lvconvert", LVConvertThinpoolCacheArgs(c.ThinpoolCachePV, c.ThinpoolLV)...); err != nil {
+			return err
 		}
-		return nil
-	})
-	return totalSize, err
+	}
+
+	return nil
 }
 
-func execCommand(cmdName string, args ...string) error {
-	if os.Getenv("RUN_WITH_SUDO") != "" {
-		cmd := exec.Command("sudo", append([]string{cmdName}, args...)...)
-		return cmd.Run()
+func LVCreateBaseArgs(vg string, basePV string) []string {
+	return []string{
+		// Create a linear LV
+		"--type", "linear",
+
+		// Name the base LV vg/base
+		"--name", "base",
+
+		// Make the base LV take up all the space on the PV
+		"--extents", "100%PVS",
+
+		// Yes, do it even if the LV already has a filesystem
+		"-y",
+
+		// Pass the volume group the base LV should be created in
+		vg,
 	}
+}
+
+func LVCreateThinpoolArgs(vg string, thinpoolPVs []string) []string {
+	lvCreateArgs := []string{
+		// Create a thin pool
+		"--type", "thin-pool",
+
+		// Name the thin pool vg/thinpool
+		"--name", "thinpool",
+
+		// Make the thin pool take up all the space on the provided PVs
+		"--extents", "100%PVS",
+
+		// Use minimum allowed chunk size for better small file efficiency
+		"--chunksize", "64k",
+
+		// Use one stripe per thinpool device to maximize performance
+		"--stripes", strconv.Itoa(len(thinpoolPVs)),
 
-	cmd := exec.Command(cmdName, args...)
-	return cmd.Run()
+		// Use a small stripe size for better IO performance on small files
+		"--stripesize", "64k",
+
+		// Pass TRIM/discard commands through to underlying storage
+		"--discards", "passdown",
```

Comment:
Hm, I don't know enough to say for sure but this seems like it'd actually worse for performance

Reply:
I think in theory yes, but from my benchmarking it didn't make a difference so I kept it to help mitigate the "out of disk space" risk.